moveit_core/kinematic_constraints/src/kinematic_constraint.cpp

```diff
@@ -1142,38 +1140,49 @@ ConstraintEvaluationResult VisibilityConstraint::decide(const moveit::core::Robo
     }
   }
 
-  shapes::Mesh* m = getVisibilityCone(state);
-  if (!m)
-    return ConstraintEvaluationResult(false, 0.0);
-
-  // add the visibility cone as an object
-  collision_env_->getWorld()->addToObject("cone", shapes::ShapeConstPtr(m), Eigen::Isometry3d::Identity());
-
-  // check for collisions between the robot and the cone
-  collision_detection::CollisionRequest req;
-  collision_detection::CollisionResult res;
-  collision_detection::AllowedCollisionMatrix acm;
-  collision_detection::DecideContactFn fn = [this](collision_detection::Contact& contact) {
-    return decideContact(contact);
-  };
-  acm.setDefaultEntry(std::string("cone"), fn);
-
-  req.contacts = true;
-  req.verbose = verbose;
-  req.max_contacts = 1;
-  collision_env_->checkRobotCollision(req, res, state, acm);
-
-  if (verbose)
+  if (target_radius_ > std::numeric_limits<double>::epsilon())
   {
-    std::stringstream ss;
-    m->print(ss);
-    RCLCPP_INFO(LOGGER, "Visibility constraint %ssatisfied. Visibility cone approximation:\n %s",
-                res.collision ? "not " : "", ss.str().c_str());
-  }
+    const auto collision_env_local = std::make_shared<collision_detection::CollisionEnvFCL>(robot_model_);
+
+    shapes::Mesh* m = getVisibilityCone(state);
+    if (!m)
+    {
+      RCLCPP_ERROR(LOGGER, "Visibility constraint is violated because we could not create the visibility cone mesh.");
+      return ConstraintEvaluationResult(false, 0.0);
+    }
 
-  collision_env_->getWorld()->removeObject("cone");
+    // add the visibility cone as an object
+    collision_env_local->getWorld()->addToObject("cone", shapes::ShapeConstPtr(m), Eigen::Isometry3d::Identity());
```

Comment:
I see why this would cause problems with multi-threading, this should work. Another way to solve this multi-threading issue is to attach the cone to the (local) robot state and check for self-collisions. That way, there shouldn't be any race conditions since there are no variable states being shared between threads. `checkCollision()` alone is thread-safe afaik. What do you think?

Reply:
I think recreating the collision environment every time we check this constraint is probably quite inefficient, so if there's a way to reduce that I think that's preferable.

> Another way to solve this multi-threading issue is to attach the cone to the (local) robot state and check for self-collisions.

I'd appreciate some pointers on how to do this -- is this using a function of the robot state or the collision environment?

An important thing to consider is that the visibility cone needs to work both when it moves with the robot and when it's fixed in the world. The way I have it written here works for both cases, but it loses some of the efficiency of the previous (attempted but buggy) approach of precalculating the transforms for fixed frames.